5-1 레스토랑 조회 구현하기

alstjr0183 opened the pull request:
안녕하세요!! 아직 reducer쪽 테스트 코드만 작성했는데 리뷰 요청드리겠습니다!!

alstjr0183 committed:
feat: redux , redux-thunk setting
redux, redux-thunk을 셋팅해주었습니다

alstjr0183 committed:
feat: index redux 셋팅
index.js에 redux를 연동해주었습니다

alstjr0183 committed:
feat: reducer.test - getRestaurantRegions 작성
레스토랑 지역을 가져오는 지역을 가져오는 reducer 테스트 코드를
작성했습니다

alstjr0183 committed:
fix: getRestaurantRegions -> getRegions로 변경
지역 리스트를 보여주는 상태인데 레스토랑이라는 부분은 굳이없어도
될거같아서 뺐습니다

alstjr0183 committed:
fix: reducer.test - getRegions it 추가
getRegions의 it을 추가한다

alstjr0183 committed:
feat: fixtures-restaurant-regions 작성
regions 테스트 상태를 작성했습니다

alstjr0183 committed:
feat: reducer-getRegions 구현

alstjr0183 committed:
fix: action , api 에서 restaurant 네이밍 제거
지역을 나타내는정보를 얻어오는 역할을하기때문에 restaurant라는 네이밍을
제거해주었습니다

alstjr0183 committed:
feat: reducer.test - getCategories 테스트 작성
reducer의 음식 카테고리를 얻는 테스트를 추가했습니다

alstjr0183 committed:
feat: fixtures-restaurant-categories 상태 추가
fixtures에 테스트 상태를 추가해주었습니다

alstjr0183 committed:
feat: categories api 로직 추가
categories 불러오는 api를 추가해주었습니다

alstjr0183 committed:
feat: actions-getCategories action 추가
카테고리 목록을 들고오는 액션을 추가했습니다

alstjr0183 committed:
feat: reducer-getCategories 작성
reducer에 카테고리 리스트를 저장시켜주는 로직을 추가하였습니다

alstjr0183 committed:
fix: reducer.test - getCategories it 설명 수정
getCategories의 설명을 수정했습니다

passwd10 commented on src/reducer.test.js:
```diff
@@ -0,0 +1,36 @@
+
+  context('state가 있는경우', () => {
+    describe('getRegions', () => {
+      it('regions를 얻는다', () => {
```
getRegions reducer는 store에 있는 regions 를 반환하는  일을 하고있어요.

alstjr0183 replied:
### 수정했습니다
역할에 맞는 설명으로 regions를 return 한다 라는 표현이 괜찮을까요?!

passwd10 commented on src/actions.js:
```diff
@@ -0,0 +1,35 @@
+
+    dispatch(getCategories(categories));
+  };
+}
```
액션함수끼리 비동기 동작을 하는 액션함수 끼리 응집해서 작성하거나, 파일로 분리해서 관리하면 유지보수하기 편하지 않을까요? ㅎㅎ

alstjr0183 replied:
### 수정했습니다
그러네요..! 저렇게 분리해주면 더 좋은거같습니다!!

passwd10 commented on src/reducer.test.js:
```diff
@@ -0,0 +1,36 @@
+    });
+  });
+
+  context('state가 있는경우', () => {
```
context를 적절하게 활용하셨어요 👍

alstjr0183 replied:
헉 감사합니다! 😄😄

passwd10 commented on src/services/api.js:
```diff
@@ -0,0 +1,15 @@
+}
+
+export async function fetchCategories() {
+  const url = 'https://eatgo-customer-api.ahastudio.com/categories';
```
중복되는 url 이 보이네요!

alstjr0183 replied:
### 수정했습니다
api 파일에 baseUrl 이라는 변수를 만들어 중복되는 부분을 제거해보았습니다!

passwd10 reviewed (commented):
먼저 올바르게 동작하는 완성된 형태의 어플리케이션을 만들고, 케이스가 하나씩 생길 때 마다 테스트를 추가해가면서 진행해보세요!

처음부터 너무 많은 케이스를 생각하고 진행하다보면 복잡도가 올라가서 구현이 힘들어질 수 있어요. 

지금 당장 사용자에게 레스토랑 목록을 보여줘야한다면, 가장 먼저 무엇을 테스트해야할까요?

alstjr0183 committed:
feat: reducer-isLoading state추가
api 호출시간동안 로딩을 표시하기위해 isLoading 상태를 추가해주었습니다

alstjr0183 committed:
fix: reducer.test - it 설명 수정
reducer.test 의 it 부분들을 역할에맞게끔 return 하는 설명으로 변경해주었습니다

alstjr0183 committed:
refactor: actions, 비동기actions 분리
actions 파일에 합쳐져있던 action들을 분리해주었습니다

alstjr0183 committed:
fix: services-api 중복되는 url 제거
baseUrl 이라는 변수를 만들어 중복을 제거해주었습니다

alstjr0183 committed:
feat: react-redux mock 추가
useDispatch , useSelector mock들을 추가해주었습니다

alstjr0183 committed:
feat: Regions.test 작성
Regions 컴포넌트의 테스트 코드를 작성하였습니다

alstjr0183 committed:
feat: Regions 구현
api 를 통해 받아온 지역 리스트를  렌더링 하는부분을 구현하였습니다

alstjr0183 committed:
fix: Regions-li key 추가
map 으로 돌린 부분에 key를 추가해주었습니다

alstjr0183 committed:
feat: app.test 작성
app.test 코드를 작성하였습니다

alstjr0183 committed:
feat: app 구현
app.test를 토대로 app을 구현하였습니다

alstjr0183 committed:
fix: app.test - context 설명 수정
app.test의 context 설명을 수정하였습니다

alstjr0183 committed:
feat: Categories.test 작성
카테고리 목록을 보여주는 테스트 코드를 작성했습니다

alstjr0183 committed:
feat: Categories 구현
카테고리를 보여주는 로직을 작성했습니다

alstjr0183 committed:
fix: reducer - api 관련 액션에 isLoading 제어 못하게끔 수정
api쪽 관련된 액션에 isloading제어권을 없애주었습니다

alstjr0183 committed:
feat: reducer.test - isLoading 상태제어 테스트 작성
isLoading을 제어할수있는 테스트코드를 작성하였습니다

alstjr0183 committed:
feat: reducer - setIsLoading 추가
reducer에 isLoading을 제어할수있는 로직을 추가했습니다

alstjr0183 committed:
feat: actions - setIsLoading 추가
actions에 isLoading을 제어하는 액션을 만들어주었습니다

alstjr0183 committed:
feat: app.test - Categories 케이스 추가
categories가 렌더링되는 테스트 코드를작성하였습니다

alstjr0183 committed:
fix: app - api 순서보장을 위한 로직수정
지역과 카테고리의 정보를 가져오는 api의 순서보장을위한 로직을 수정했습니다

alstjr0183 committed:
fix: app.test 테스트통과를 위한 기본값 추가
Regions , Categories의 가져오는값에 기본값을 설정했습니다

alstjr0183 commented:
안녕하세요 선생님..! 말씀주신부분은 수정했습니다
### 질문있습니다!

app 테스트코드에서 테스트 돌리면 이런 오류가 나길래
<img width="538" alt="image" src="https://user-images.githubusercontent.com/57708817/177544048-d4586d73-7689-413f-b9ea-186dd433b65c.png">


이런식으로 기본값 설정해주었더니 에러가 안뜨는데 이유가 뭘까요?! reducer쪽에서는 categories 기본값이 설정되어있는데 따로 저렇게 해주어야하나요?
<img width="491" alt="image" src="https://user-images.githubusercontent.com/57708817/177544246-d939aa85-03bc-4ef1-b2a5-42bbbbdc4830.png">

passwd10 commented on src/services/api.js:
```diff
@@ -1,13 +1,15 @@
+const baseUrl = 'https://eatgo-customer-api.ahastudio.com';
```
잘 분리하셨어요! 상수는 대문자 snakecase로 표현한답니다 ㅎㅎ

alstjr0183 replied:
### 수정했습니다
상수들은 대문자로 표시해주는걸로 약속을 하면 더 파악하기가 쉽겠네요!

passwd10 commented on src/App.test.jsx:
```diff
@@ -0,0 +1,42 @@
+
+  context('api 호출전', () => {
+    it('로딩중입니다! 가 렌더링된다', () => {
+      useSelector.mockImplementation((selector) => selector({
```
useSelector도 useDispatch처럼 describe 하위에서 한번만 선언해서 사용하시면 될 것 같아요! 테스트마다 selector 값을 다르게 주고싶으셔서 테스트 내부에서 선언하신 것 같아요. 이런 경우엔 `given2` 라는 라이브러리를 사용해서 lazy 하게 변수를 할당할 수 있어요!

alstjr0183 replied:
아직 수정중입니다 given2 사용방법을 찾아보고있습니다.!! ㅠ

passwd10 commented on src/App.jsx:
```diff
@@ -1,5 +1,27 @@
+    dispatch(loadRegions());
+  }, []);
+
+  if (isLoading) {
+    return <div>로딩중입니다!</div>;
+  }
```
로딩에 대한 처리를 까지 잘 해주셨네요 👍

alstjr0183 replied:
감사합니다!!!!!

passwd10 commented on src/App.jsx:
```diff
@@ -1,5 +1,27 @@
+import React, { useEffect } from 'react';
```
React 를 import 하지 않아도 될 것 같아요.

alstjr0183 replied:
넵! 수정했습니다!!

passwd10 commented on src/reducer.test.js:
```diff
@@ -1,10 +1,16 @@
 import reducer from './reducer';
 
 describe('reducer', () => {
+  it('isLoading 상태를 변경한다', () => {
```
다른 reducer 테스트와 같은 형태로 작성해주세요!

```javascript
describe('setIsLoading', () => {
  ...
})
```

alstjr0183 replied:
### 수정했습니다
앗 그러네요 통일감있게 작성해야겠습니다!

passwd10 commented on src/App.test.jsx:
```diff
@@ -38,5 +38,15 @@ describe('App', () => {
+
+      const { getByText } = renderApp();
+
+      expect(getByText(categories[0].name)).not.toBeNull();
```
```javascript
expect(container).toHaveTextContent(categories[0].name);
```

다음과 같이 표현할수도 있을 것 같아요

alstjr0183 replied:
### 수정하였습니다
변경 전 보다 흐름상 읽기가 더편해진거같네요!

passwd10 commented on src/App.jsx:
```diff
@@ -2,17 +2,25 @@ import React, { useEffect } from 'react';
 
   const { isLoading } = useSelector((state) => state);
 
+  const loadData = async () => {
+    await dispatch(loadRegions());
+    await dispatch(loadCategories());
```
async/await 을 사용하신 이유가 있을까요? async/await 은 어떤경우에 사용하는 함수일까요?

alstjr0183 replied:
### 질문있습니다!

api 통신하는 부분이 비동기로 작동하니까 api 호출이 끝나면 isLoading을 닫아주기 위해서 순서보장을 위해 사용했었는데 좀더 좋은방법이 있을까요?!

현재까지는 3가지 방법을 사용해봤는대요
### 1. reducer 쪽 api 호출 2개의 관련로직에 isLoading : false 추가
 api 호출후 isLoading을 닫아주려는 시도를해봤는대 하나만 api호출에성공해도 닫히기 때문에 이방법을 패스했습니다!
![image](https://user-images.githubusercontent.com/57708817/177662090-46535281-0ea7-405a-a792-4093d3328249.png)

### 2. 로직한곳 에만 isLoading을 주는방법
이방법은 api 통신이 비동기라 둘중에 어느쪽이 먼저 빨리 끝날지 랜덤이라 패스했었습니다!
![image](https://user-images.githubusercontent.com/57708817/177661794-9f892a94-1259-4e43-b78e-83967b8eee4c.png)

### 3. actions-async 에서 2가지 api를 같이 호출하는 로직을 만드는방법
이렇게 합치면 구현이 되긴하는데 이 api들을 다른곳에서 또 사용한다했을때 범용성이 떨어지는거같더라구요! 그래서 패스했습니다
![image](https://user-images.githubusercontent.com/57708817/177662422-6a1a647c-ebe1-49bd-a7d8-f9e58023f0aa.png)

passwd10 replied:
regions 전용 isLoading과 categories 전용 isLoading을 각각 만들어서 관리해도 되지 않을까요?

비동기 action 내부에서 loading 상태를 dispatch하면 외부에서 동기 비동기를 신경쓰지 않을 수 있을 것 같아요!

alstjr0183 replied:
### 수정했습니다
헉 그러네요.. 그생각을 못했었네요!! 감사합니다

passwd10 reviewed (commented):
`regions가 렌더링된다` 테스트에서 categories 를 mocking하지 않고 있어서 나오는 에러같아요. categories 도 useSelector mocking 해주세요!

alstjr0183 committed:
fix: app - import react 제거
의미없는 import 를 제거하였습니다

alstjr0183 committed:
fix: reducer.test - setIsLoading 설명 통일감있게 변경
setIsLoading도 다른 테스트와 같은형태로 변경해주었습니다

alstjr0183 committed:
fix: app - useEffect import
react import만 제거해야하는데 실수로 useEffect도 제거해버렸네요
복구했습니다!

alstjr0183 committed:
refactor: services - api baseurl 대문자로 변경
baseurl은 상수이니 대문자로 변경했습니다

alstjr0183 committed:
fix: app.test - expect 방법 변경
expect 부분의 container의 toHaveTextContent로 변경하였습니다

alstjr0183 committed:
fix: Regions, Categories 기본값제거
테스트 통과를 위해 임시방편으로했던 기본값을 제거했습니다

alstjr0183 committed:
fix: app.test - useSelector mocking 추가
regions , categories mocking 추가

alstjr0183 commented:
안녕하세요 선생님 리뷰 주신 내용중에선 given2 사용방법을 아직 찾아보고있습니다..!

아 선생님..! 이렇게 하니까 통과는 되는데 이렇게 해두 괜찮은건가요?..
저 부분을 통합해서 테스트를 하는게 좋을까요?!
<img width="535" alt="image" src="https://user-images.githubusercontent.com/57708817/177771567-0e965c12-ed87-4b4c-b9c7-ccff62e9cbb2.png">

alstjr0183 committed:
feat: give2 라이브러리 설치

alstjr0183 committed:
feat: reducer.test - setRegion 추가 , fixtures - region 추가
reducer.test 지역을 체크하는 로직을 추가해주었습니다
fixtures 에 테스트로 사용할 변수를 추가했습니다

alstjr0183 committed:
feat: reducer - setRegion 구현
reducer 쪽에 setRegion을 구현했습니다

alstjr0183 committed:
feat: actions - setRegion 구현
지역을 체크하는 액션을 만들어주었습니다

alstjr0183 committed:
feat: RegionsContainer.test 작성
이벤트와 , ui 분리를 위하여 container를 작성해주었습니다

alstjr0183 committed:
feat: RegionsContainer 구현
이벤트, ui 분리를 위해 container를 만들었습니다

alstjr0183 committed:
fix: RegionsContainer 추가로 인한 로직 수정
ui , 이벤트 분리를 위한 작업을 해주었습니다

alstjr0183 committed:
fix: app - Regions 를 RegionsContainer 로 변경
app 쪽에 사용하는 컴포넌트를 변경해주었습니다

alstjr0183 committed:
feat: Regions - click 이벤트 적용
button 을 클릭하면 지역이 선택되는 이벤트를 적용시켜주었습니다

alstjr0183 committed:
fix: RegionsContainer - handleClickSelectRegion 수정
handleClickSelectRegion 이벤트에서  e.target.value 를받아왔는데 input쪽에 name을 추가한후 e.target.name 으로 변경해주었습니다

alstjr0183 committed:
fix: RegionsContainer.test 설명 수정
잘못된 설명을 수정했습니다

alstjr0183 committed:
fix: categories.test , regions.test 테스트 수정
테스트를 좀더 직관적으로 변경했습니다

alstjr0183 committed:
feat: RegionsContainer.test - click event 테스트 작성
지역을 누르면 이벤트가 발동되는 테스트 코드를 작성했습니다

alstjr0183 committed:
feat: Regions.test - 렌더링 케이스 추가
지역이 선택되었을때 선택된 지역에 (V) 가 뜨는 케이스를 작성해주었습니다

alstjr0183 committed:
feat: Regions - 클릭이벤트 후 구현
지역을 선택했을때 (V) 가 뜨게끔 구현해주었습니다

alstjr0183 committed:
feat: 모든 테스트케이스 mockClear 추가
모든 테스트 케이스에 clearAllMocks 를 추가해주었습니다

passwd10 commented on src/RegionsContainer.test.jsx:
```diff
@@ -0,0 +1,32 @@
+    fireEvent.click(getByText('서울'));
+
+    expect(dispathch).toBeCalled();
+  });
```
작성하신 테스트 시나리오와 실제 확인하는 테스트가 다르네요!

alstjr0183 replied:
그 부분은 현재 이렇게 바껴있습니다!!
![image](https://user-images.githubusercontent.com/57708817/177888851-45361e98-7f9d-4d5a-abd1-63560074b495.png)

passwd10 commented on src/Regions.jsx:
```diff
@@ -3,7 +3,7 @@ export default function Regions({ regions, onClickSelectRegion }) {
     <ul>
       {regions.map((i) => (
```
i 는 아무 의미없는 변수에요. 의미를 담은 네이밍을 고민해보세요!

alstjr0183 replied:
region , regions 이렇게 변경하려했는데 다른곳에서 사용중인 네이밍이라 구조분해할당으로해서 { name, id  }를 사용했는데 괜찮을까요?

passwd10 commented on src/RegionsContainer.jsx:
```diff
@@ -10,7 +10,7 @@ export default function RegionsContainer() {
 
   const handleClickSelectRegion = (e) => {
-    dispatch(setRegion(e.target.value));
+    dispatch(setRegion(e.target.name));
```
target의 name을 가져오는 이유를 알 수 있을까요?

alstjr0183 replied:
button쪽에서 e.target.value를 하니 아무값도 안나오더라구요! 그래서 Input쪽에 name을 주고 e.target.name을 들고와 사용했습니다!

passwd10 commented on src/Regions.test.jsx:
```diff
@@ -1,27 +1,35 @@
-
-  const renderRegions = () => render((
-    <Regions regions={regions} onClickSelectRegion={handleClickSelectRegion} />
+  const renderRegions = (paramRegion) => render((
```
param 이라는 prefix 는 불필요해보여요 ㅎㅎ

alstjr0183 replied:
region으로 네이밍을 주려했는데 다른곳에서 사용중이더라구요..! 이런 경우가 자주있는데 선생님께선 보통 어떤 네이밍을 주시나요?!

passwd10 commented on src/Regions.test.jsx:
```diff
@@ -4,6 +4,10 @@ import { region, regions } from './fixtures/restaurant';
 
 import Regions from './Regions';
 
+beforeEach(() => {
+  jest.clearAllMocks();
+});
+
```
describe 하위에서 선언해주시면 테스트가 더 자연스럽게 읽힐 것 같아요!

alstjr0183 replied:
그러네요!! 앞으로는 describe안에 넣겠습니다!

passwd10 reviewed (commented):
happy path에 대한 테스트를 잘 작성하셨어요. 👍 

만약 fetch한 categories가 undefined 인 경우엔 어떻게 해야할까요?

예외처리도 테스트를 통해 검증해보시고 어떻게하면 해결 할 수 있을지 고민해보시면 좋을 것 같아요!

alstjr0183 committed:
fix: map 함수 구조분해할당해서 사용
map함수를 사용할때 구조분해할당을 사용했습니다

alstjr0183 committed:
fix: mock clear 사용중인 test 수정
읽기편하게 describe 안에 넣어주었습니다

alstjr0183 committed:
fix: app - api 호출 로직수정
api 호출을 하는 각각의 액션에 따로 isLoading을 주어 관리하게끔 수정했습니다

alstjr0183 commented:
시간이부족해 리뷰받은 내용들만 작업했네요..! 진도를 더나가겠습니다!

alstjr0183 committed:
feat: reducer.test - setCategory 작성
카테고리를 선택하는 리듀서 테스트를 작성했습니다

alstjr0183 committed:
feat: actions - setCategory , fixtures - category 추가
action과 테스트변수 값을 추가했습니다

alstjr0183 committed:
feat: reducer - setCategory 구현
카테고리 선택하는 로직을 구현했습니다

alstjr0183 committed:
fixures: fixtures - restaurnt - cateory 추가
테스트 데이터에 category를 추가했습니다

alstjr0183 committed:
feat: categoriesContainer 테스트 작성
테스트를 작성했습니다

alstjr0183 committed:
feat: CategoriesContainer 구현
작성한 테스트를 토대로 구현했습니다

alstjr0183 committed:
fix: Categories.test 수정
CategoriesContainer 추가로인한 수정

alstjr0183 committed:
fix: Categories , app 수정
로직변경으로인한 컴포넌트 수정

alstjr0183 committed:
feat: reducer.test - getRestaurants 작성
레스토랑 리스트를 들고오는 테스트를 작성했습니다

alstjr0183 committed:
feat: test 데이터 추가
fixtures - resturant - restaurants 테스트 데이터를 추가했습니다

alstjr0183 committed:
feat: actions - getRestaurants 추가
레스토랑 리스트를 받아오는 액션을 추가했습니다

alstjr0183 committed:
feat: reducer - getRestaurants 구현
레스토랑 리스트를 가져오는 로직을 구현했습니다

alstjr0183 committed:
fix: fixtures수정
category -> categoryId로 수정했습니다

alstjr0183 committed:
fix: category -> categoryId 네이밍 변경

alstjr0183 committed:
fix: category 상태 id로 저장
category 를 e.target.name으로 저장해주었는데 id로 변경해주었습니다

alstjr0183 committed:
fix: category 사용한 부분수정
categoryId로 변경해주었습니다

alstjr0183 committed:
feat: actions - loadRestaurants , fetchRestaurant 추가

alstjr0183 committed:
fix: actions-async loadRestaurants 수정
파라미터를받아서 fetchRestaurans에 넘겨주었습니다

alstjr0183 committed:
feat: reducer, actions - setStartLoading
로딩을 true로 만들어주는 로직을 구현했습니다

passwd10 commented on src/CategoriesContainer.test.jsx:
```diff
@@ -0,0 +1,36 @@
+    jest.clearAllMocks();
+  });
+
+  const dispathch = jest.fn();
```
오타가 있네요 ㅎㅎ

alstjr0183 replied:
### 수정했습니다!

alstjr0183 committed:
feat: app.test 수정
restaurants 추가로인한 테스트수정

alstjr0183 committed:
feat: 레스토랑 리스트 api 구현

alstjr0183 commented:
안녕하세요 선생님..! 피드백주신 내용들 수정해보았습니다
마지막에 지역이랑 카테고리 선택했을때 레스토랑 리스트들 나오는 부분은 너무너무헷갈려서 구현부터 해버렸습니다 ㅠㅠ
resturants 테스트 코드는 좀더 고민후 커밋하겠습니다..!

passwd10 commented on src/reducer.js:
```diff
@@ -2,9 +2,11 @@ const initialState = {
   isLoading: {
     regions: true,
     categories: true,
+    restaurants: false,
```
로딩 초기상태가 다 다른 이유가 있을까요?

alstjr0183 replied:
regions랑 categories는 페이지가 렌더링되자마자 api호출을해야하는 부분이라서true로 했는데
restaurants는 지역이랑 카테고리가정해지면 불러와야해서 false로 했었습니다..!

passwd10 commented on src/Categories.jsx:
```diff
@@ -3,9 +3,16 @@ export default function Categories({ category, categories, onClickSelectCategory
         <li key={id}>
-          <button onClick={onClickSelectCategory} name={name} type="button" key={id}>
+          <button
+            onClick={() => {
+              onClickSelectCategory(id);
+            }}
```
저는 인라인에서 바로 함수를 작성하기보단, 핸들러 함수를 따로 선언해서 넣어주는 편이에요. 함수명만 읽고도 대략 어떤역할을 하는지 알 수 있어서 가독성이 올라가거든요!

민석님은 어떻게 생각하시나요? 정답이 있는건 아니라서 의견을 자유롭게 나눠보면 좋을 것 같아요ㅎㅎ

alstjr0183 replied:
저도 pass님 처럼 생각하는데요! 저기서 id를 넘겨줘야하는데 onClick={onCLickSelectCategory} 이 형태로 id 넘겨줄수있는 방법이있나요?!

passwd10 replied:
id 를 인자로 받는 핸들러 함수를 만들면 될 것 같아요 ㅎㅎ

currying 함수로 작성해보세요!

https://ko.javascript.info/currying-partials

alstjr0183 committed:
fix: 오타수정
dispatch 오타를 수정했습니다

alstjr0183 commented:
안녕하세요! 선생님 한주동안 리뷰 감사합니다 ㅎㅎ
앞으로도 잘부탁드릴게요!!
